clang/lib/Sema/SemaOverload.cpp

```diff
@@ -10243,6 +10260,15 @@ static bool checkAddressOfFunctionIsAvailable(Sema &S, const FunctionDecl *FD,
                                               bool Complain,
                                               bool InOverloadResolution,
                                               SourceLocation Loc) {
+  if (Complain && S.getLangOpts().SYCLIsDevice &&
+      S.getLangOpts().SYCLAllowFuncPtr) {
```

Comment:
According to the SYCL spec, it is not allowed to perform a call through a function pointer, but strictly speaking, it is allowed to take an address of a function. Moreover, there are cases when function pointers are used, but in so straightforward manner so they can easily be deduced by the compiler and transformed into direct calls; another example is when function pointer is declared as `constexpr` - for the sake of better adoption of DPC++/SYCL there is a desire to support such function pointers at some point (once corresponding spec clarifications will be made).

However, we have an undocumented extension, which allows to use truly dynamic function pointers and the tricky part there is that compiler can't always understand which functions should be included into device code and which are not - for example if the function is only used indirectly through a function pointer and call site and definition are in different translation unit. In that case we can encounter linker error at later stages and for truly dynamic function pointers it would be great to have a diagnostic for such situations.

Having in mind that the status of `constexpr` function pointers is still unknown, we have decided to implement diagnostic _only_ under the flag as it allows us:
- not to break existing codes where function pointer is taken, but not used
- have a diagnostic where function pointer is taken in order to be actually used later

The flag serves as indicator that function pointers are actually going to be used, because user explicitly opt-ins for extra features

Reply:
Thanks for the explanation!